hadoop-hdds/docs/content/design/decommissioning.md

```diff
@@ -0,0 +1,720 @@
+
+### In-flight examples
+
+#### Maintenance + inflight
+
+  Node with replica | Node status
+  ------------------|------------
+  A                 | HEALTHY
+  B                 | ENTERING_MAINTENANCE
+  C                 | (missing) + inflight
+
+
+  Counter            | Value
+  ------------------ | -------------
+  expectedCount      | 3
+  healthy            | 1
+  maintenance        | 1
+  in-flight          | 1
+  replicaCount       | 0
+
+Here one have one node ENTERING_MAINTENANCE state, and one replica is missing and already started to be replicated. We don't need to start a new copy and node B can be moved to the IN_MAINTENANCE mode.
+
+```
+Replica Count (0) = expectedCount(3) - (healthy(1) + maintenance(1) - inFlihgt(1))
+```
+
+
+#### Maintenance + inflight
+
+  Node with replica | Node status
+  ------------------|------------
+  A                 | DECOMMISSIONING
+  B                 | (missing) + inflight
+  C                 | (missing) + inflight
+
+
+  Counter            | Value
+  ------------------ | -------------
+  expectedCount      | 3
+  healthy            | 0
+  maintenance        | 0
+  in-flight          | 2
+  replicaCount       | 1
+
+We have on emissing replicas, but and node A can not be DECOMMISSIONED as we have no HEALTHY replicas at all.
+
+
+## Statefulness
+
+SCM stores all the node state in-memory. After a restart on the SCM side the datanode state can be lost.
+
+**Ozone doesn't guarantee that decommissioning/maintenance mode state survives the SCM restarts!!!**
+
+ * If SCM restarts DECOMMISSIONED nodes will not report any more container reports and the nodes won't be registered.
+ * ENTERING_MAINTENANCE and DECOMMISSIONING nodes will became HEALTHY again and the decommissioning CLI command should be repeated.
+ *  IN_MAINTENANCE nodes will become DEAD and all the containers will be replicated.
+
+ *Ozone assumes that the maintenance mode is used short-term and SCM is not restarted during this specific period.*
+
+
+*Reasoning*:
+
+Neither of the node state nor the container state are persisted in SCM side. The decommissioned state can be stored on the SCM side (or on the SCM side and the datanode side) which can provide better user experience (and may be implemented).
+
+But to support maintenance mode after restart all the container information is required to be persisted (which is a too big architectural change).
+
+To make a replication decision replication manager needs the number of healthy replicas (they are reported via heartbeats) AND the number of containers on the node which is in maintenance mode. The later one is not available if the SCM is restarted as the container map exists only in the memory and the node which is turned off can't report any more container reports. Therefore the information about the existing containers on the node which is in the maintenance mode **can't be available'**. 
+
+## Throttling
+
```

Comment:
How does SCM currently limit replication in the case of dead nodes? Possibly the same mechanism can be used for decommissioning?

Reply:
Yes, @anuengineer and @nandakumar131 commented the same. I rewrote this section in the latest version just to mention that throttling is possible but independent from the decommissioning and should be discussed in a separated document. @nandakumar131 mentioned that he would like to create a separated doc about throttling...